File: sdk/ts/src/noise/cipherState.ts
import { aeadDecrypt, aeadEncrypt } from './primitives.js';

/** Noise's CipherState: (k, n). Encrypts/decrypts one direction of one session. */
export class CipherState {
  private key: Uint8Array | null = null;
  private n = 0n;

  initializeKey(key: Uint8Array): void {
    this.key = key;
    this.n = 0n;
  }

  hasKey(): boolean {
    return this.key !== null;
  }

  encryptWithAd(ad: Uint8Array, plaintext: Uint8Array): Uint8Array {
    if (this.key === null) return plaintext;
    const ciphertext = aeadEncrypt(this.key, this.n, ad, plaintext);
    this.n += 1n;
    return ciphertext;
  }

  decryptWithAd(ad: Uint8Array, ciphertext: Uint8Array): Uint8Array {
    if (this.key === null) return ciphertext;
    const plaintext = aeadDecrypt(this.key, this.n, ad, ciphertext);
    this.n += 1n;
    return plaintext;
  }
}

File: sdk/cpp/src/noise/cipher_state.cpp
#include "noise/cipher_state.hpp"

namespace relayly::noise {

void CipherState::InitializeKey(const Key& key) {
  key_ = key;
  n_ = 0;
}

std::vector<std::uint8_t> CipherState::EncryptWithAd(const std::vector<std::uint8_t>& ad,
                                                     const std::vector<std::uint8_t>& plaintext) {
  if (!key_.has_value()) {
    return plaintext;
  }
  auto out = AeadEncrypt(*key_, n_, ad, plaintext);
  ++n_;
  return out;
}

std::vector<std::uint8_t> CipherState::DecryptWithAd(const std::vector<std::uint8_t>& ad,
                                                     const std::vector<std::uint8_t>& ciphertext) {
  if (!key_.has_value()) {
    return ciphertext;
  }
  auto out = AeadDecrypt(*key_, n_, ad, ciphertext);
  ++n_;
  return out;
}

}  // namespace relayly::noise
